### ISSUE TITLE
POC: Evaluate optimizing disk i/o to improve overall process execution latency and throughput

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#11331]
**Description**

Related to https://github.com/camunda/product-hub/issues/642

Disk i/o is currently a significant contributor to processing latency. Flushing the committed records to the disk is a blocking i/o and raft thread is blocked during this time. We have seen that disabling flush improves performance, but it comes with safety issues as the unflushed data can be lost after a node crash. The goal of this POC is to evaluate how to asynchronously flush events to the log without compromising safety to improve the overall latency and throughput.

**Comment by deepthidevaki** [comment]
Potential concepts to evaluate:

1. Flush periodically (every X ms) both on leader and follower. But commit without flushing.
   This is not a "safe" solution. It has the same problems as disabling flush. The only difference is that the max period of inconsistency is limited to configured X ms. We are evaluating this to get an understanding of the maximum performance possible.
2. Flush periodically (every X ms or every Y bytes ). Commit only after flushing.
   This is the "safe" solution.  But delaying the commit can increase the processing latency. This requires changes in the raft communication protocol
      - Follower should ack receiving AppendRequest before flushing, so that the leader can send more events.
      - Follower must inform leader when it flushes.
      - The leader should not commit immediately when the follower ack, but should wait until the follower flushes.
3.  Move journal to another thread. Writes and flush happen on that thread. 
     The goal is not to block the raft thread while blocking disk i/o. A record is committed only after it is flushed. Having a second thread can lead to concurrency issues. We can look into it if the poc shows there is a significant performance improvement with this approach.

**Comment by Zelldon** [comment]
4. https://github.com/camunda/zeebe/issues/6044

**Comment by deepthidevaki** [comment]
> 4. [Make state directory configurable  #6044](https://github.com/camunda/zeebe/issues/6044)

Thanks @Zelldon  In cloud, we could mount a local disk to runtime. Might be possible to try it out in hackweek, if other POCs are done.

**Comment by npepinpe** [comment]
Regarding 4, I think one interesting use case is different disks for different partitions to avoid maxing out IOPS. Though you would hope all partitions receive equal load, I'm not so sure about it :smile: 

I would also add:

5. Add bindings to [madvise](https://man7.org/linux/man-pages/man2/madvise.2.html). Informing the OS that we access things sequentially (`MADV_SEQUENTIAL`) provided a noticeable improvement for such a small change. Additionally, we could look into advising the OS as well about `MADV_DONOTNEED` when no readers/writers are opened for a segment (to free those pages and not the ones we may need), and `MADV_WILLNEED` after a seek (to tell the OS to start loading pages from the seek onwards). That said, that was tested with 128MB segments, perhaps it's not a big deal on smaller ones. But it's a cheap addition, imo, with obvious benefits (at least just the `MADV_SEQUENTIAL` anyway).

**Comment by npepinpe** [comment]
I did a quick test of using [copy_file_range](https://man7.org/linux/man-pages/man2/copy_file_range.2.html) (with standard JDK `FileChannel#transferTo`, which should use [sendfile](https://man7.org/linux/man-pages/man2/sendfile.2.html) on Linux).

> **Note** In theory, OpenJDK should also use `copy_file_range` on Linux if it's available, but on my system it doesn't (even if my system supports `copy_file_range`). So in my PR I implemented it manually, but we could also remove it and just use `transferTo` and rely on the built-in optimization if it works as expected for most people. Maybe there's something weird going on with my setup.

I also added a metric to track the time spent pre-allocating vs just creation, so we can properly evaluate different strategies (thanks for the hint @deepthidevaki). Turns out pre-allocation is only about 50% of the time spent creating the segment. This is regardless of the strategy used to pre-allocate.

Anyway, here are the results for just pre-allocation:

| Quantile | COPY | FILL |
|----------|------|------|
| p50 |  200-240ms  |  240-360ms  |
| p90 |  2.12-3.58s |  2.81-4.75s |
| p99 |  2.73-4.23s |  3.73-4.97s |

> **Note** As usual I ran benchmarks on isolated nodes, with 1350m CPUs - otherwise all values are the same as our standard benchmarks.

So the `COPY` approach (the PoC) seems slightly better than the `FILL` approach (our current default one). That said, I couldn't test the `copy_file_range` optimization since our cloud nodes use Linux 5.10.x, and `copy_file_range` is available only on 5.13+. Similarly, I couldn't find documentation on ext4 support for copying via reflink (i.e. making this operation a copy-on-write one - for example, on my system which is brtfs, `copy_file_range` copies a 128MB file in 40ns, without noticeable degradation when it comes to writing), so I don't know if it will be fast in our SaaS.

Still, it seems like an easy pick to use `transferTo`, even at the cost of maintaining an "extra" segment (i.e. higher disk usage).

**Comment by deepthidevaki** [comment]
I ran a few benchmarks with the following configurations:

1. baseline from @npepinpe [branch with madvice](https://github.com/camunda/zeebe/commit/b8b16b33aa768f5efc5a16d7581ecc867017ac03)
  * set MADV_SEQUENTIAL 
  * default flush
2. baseline + explicit flush disabled
3. baseline + flush on another thread [branch](https://github.com/camunda/zeebe/commit/f9dd4e1debaa19ee4d78f13a22d1d34290ee8722)
    - do not wait for flush to complete before commit or ack.
    - This is not safe as we do not ensure the record is flushed before committing. But it would give us an estimate of performance with this approach.
4. 3 + add an artificial delay of 10ms between every flush [branch](https://github.com/camunda/zeebe/commit/e795ac653a48083139eaf771656d76704b3a957f)
5. 4 + specify the address range of the buffer to flush [branch](https://github.com/camunda/zeebe/commit/9cf12e5bc846b7ee8f6ca42cf200222c63da8085)
    - @Zelldon asked if OS would unnecessary flush the regions of the buffer that are not changed. With this change we want to evaluate if we can improve it, if OS is not taking care of it. The assumption is that we can prevent unnecessary flush of the regions of the buffer that are not changed. 

Results:
p50 process instance duration
![image](https://user-images.githubusercontent.com/1997478/210586848-a4399f21-e8e7-47e6-b5d1-a2c6bc610a69.png)

p50 record processing latency (duration between a record is written to the dispatcher until it is processed.)
![image](https://user-images.githubusercontent.com/1997478/210587133-7ff439a2-12f5-4d4c-bdb8-3c88a78bed61.png)

p99 record processing latency
![image](https://user-images.githubusercontent.com/1997478/210588828-dc21a03d-8a4f-4a68-9d8d-3ba1d6377de8.png)

### TLDR 
- Moving flush to another thread did not have any noticeable impact
- Adding 10ms delay contributed to a small improvement
   - Number of flushes per second reduced to almost half. Although there is no noticeable difference to Written bytes/s. 
- Specifying address range in the buffer to flush contributed to a small improvement.

Overall, delaying flush by 10ms together with specifying flush (configuration 5) seems to be winning. It is not yet clear how much of this is actual improvements vs influence from the environment. But it seems promising to look into this area further. We still have to evaluate if we get similar improvement if we wait for the flush to complete before committing.

**Comment by npepinpe** [comment]
Do we know what is the max IOPS and write throughput we can get with our disks right now? Are we anywhere close in our benchmarks?

**Comment by deepthidevaki** [comment]
> Do we know what is the max IOPS and write throughput we can get with our disks right now? Are we anywhere close in our benchmarks?

Last time I checked, we were not reaching the documented limits for throughput.

**Comment by npepinpe** [comment]
From what I understand, we have 0.48 MB/s per GB (disk size). So for a 128GB disk, we have a max of 61.44 MB/s, though we usually cap out between 40-50 MB/s.

**Comment by Zelldon** [comment]
Do you use the recently update config for Benchmarks? Then you have only 32 gig disk.

**Comment by npepinpe** [comment]
No, I used whatever is on the branch I created some time in Dec I guess :smile:

**Comment by deepthidevaki** [comment]
> From what I understand, we have 0.48 MB/s per GB (disk size). So for a 128GB disk, we have a max of 61.44 MB/s, though we usually cap out between 40-50 MB/s.

I was running with a lower load. So metrics reported shows upto only 25 MB/s.

> Do you use the recently update config for Benchmarks? Then you have only 32 gig disk.

I changed it to 64Gi.

**Comment by deepthidevaki** [comment]
With 64Gi, I observed 40MB/s when running with a higher starter rate.

**Comment by npepinpe** [comment]
Strange, it should cap around 30 MB/s. But I don't know if their docs report hard caps or just a minimum guaranteed bound :shrug:

**Comment by npepinpe** [comment]
It would be cool if we could estimate how many bytes we write per flush. I guess it becomes more difficult with async flush :thinking: But I'd be curious, because based on this we could try flushing every X MB/s instead (though I suspect that value highly depends on the disk throughput you get).

BTW, can you commit the changes you made (the most promising ones)? Or push them to some branch.

**Comment by deepthidevaki** [comment]
> It would be cool if we could estimate how many bytes we write per flush. I guess it becomes more difficult with async flush thinking But I'd be curious, because based on this we could try flushing every X MB/s instead (though I suspect that value highly depends on the disk throughput you get).

We can add some metrics for it. Not sure if there are other ways to monitor it.

> BTW, can you commit the changes you made (the most promising ones)? Or push them to some branch.

They are already available in a branch. It is linked in the above commit.

**Comment by deepthidevaki** [comment]
@Zelldon I did another experiment to evaluate whether limiting the range of buffer to be flushed by using `MappedByteBuffer.force(int, int)` would help. In this the only change is related to it. So we can see if this has any impact of this in isolation without other changes.

There is no noticeable improvement.

![image](https://user-images.githubusercontent.com/1997478/210809792-bf999b75-5e9d-4248-be72-6c7d5ca41358.png)

[Relevant commit](https://github.com/camunda/zeebe/commit/9d29d3519c9b3d895f877bfec2223abe75a4bce2)

**Comment by deepthidevaki** [comment]
No. Commit latency, IOPS, flush latency all seems to be similar.

**Comment by deepthidevaki** [comment]
Update:
We observed that this line has more impact on the performance than flush. Just commenting out this line https://github.com/camunda/zeebe/blob/main/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java#L1041  gave a significant performance improvement comparable to disabling flush. 
![image](https://user-images.githubusercontent.com/1997478/211001923-0b1f4241-3cde-4402-9a8a-73a3c74be97a.png)
More discussions in [Slack](https://camunda.slack.com/archives/C04HYU78KCG/p1672938166280889)

I think we should re-evaluate the above ideas (moving flush to another thread) again while considering how to take setting lastWrittenIndex out of hot path.

**Comment by npepinpe** [comment]
Great catch! Ideally, we would find a way to combine both flushes. When we write to the meta store, we have to flush 4kb at least regardless of the size. If we could figure out a way to have this information part of the normal write path such that flushing it goes along with the rest of our append data :thinking_face:

**Comment by deepthidevaki** [comment]
Last week we tried out moving flush out of hot path
Summary:
-  Moving flush and metastore update to another thread improves latency. But when we delay commit until flush is completed, the latency is similar to the baseline.
- Flush and metastore update on another thread, and flush on every X bytes reduces the IOPS significantly. But when we delay commit until flush is completed, latency is worse than the baseline.
- We also evaluated batch processing + no metastore update with the C8 benchmark setup. Removing metastore update did not yield significant latency improvements. So we did not evaluate any optimizations to metastore update. 

We were not able to improve latency while keeping safety guarantees similar to sync flush. Alternative would be to provide a trade-off between safety and performance by allowing users to configure how often to flush.
